upf/upf_pfcp_api.c

```diff
@@ -687,6 +731,66 @@ validate_ue_ip (upf_session_t * sx, upf_nwi_t * nwi,
     response->failed_rule_id.id = pdr->pdr_id;				\
   } while (0)
 
+upf_nat_addr_t *
+get_nat_addr (upf_nat_pool_t * np)
+{
+  upf_nat_addr_t *this_addr = NULL, *addr = NULL;
+  u32 least_locked_ports = ~0;
+  vec_foreach (this_addr, np->addresses)
+  {
+    if ((this_addr->used_blocks < least_locked_ports)
+	&& (this_addr->used_blocks < np->max_blocks_per_addr))
+      {
+	least_locked_ports = this_addr->used_blocks;
+	addr = this_addr;
+      }
+  }
+  return addr;
+}
+
+int
+upf_alloc_and_assign_nat_binding (upf_nat_pool_t * np, upf_nat_addr_t * addr,
+				  ip4_address_t user_ip, upf_session_t * sx,
+				  pfcp_tp_created_binding_t * created_binding)
+{
+  u16 port_start, port_end;
+  int err = 0;
+
+  port_start = np->min_port;
+  port_end = port_start + np->port_block_size;
+
+  upf_nat_create_binding =
+    vlib_get_plugin_symbol ("nat_plugin.so", "nat_create_binding");
+
+  do
+    {
+      err =
+	upf_nat_create_binding (user_ip, addr->ext_addr, port_start, port_end,
+				np->vrf_id);
+      if (!err)
+	{
+	  addr->used_blocks += 1;
+	  sx->nat_addr = addr;
+	  created_binding->block = vec_dup (np->name);
+	  created_binding->outside_addr.as_u32 = addr->ext_addr.as_u32;
+	  created_binding->port_range.start_port = port_start;
+	  created_binding->port_range.end_port = port_end;
+	  SET_BIT (created_binding->grp.fields,
+		   TP_CREATED_BINDING_NAT_PORT_BLOCK);
+	  SET_BIT (created_binding->grp.fields,
+		   TP_CREATED_BINDING_NAT_OUTSIDE_ADDRESS);
+	  SET_BIT (created_binding->grp.fields,
+		   TP_CREATED_BINDING_NAT_EXTERNAL_PORT_RANGE);
+	  return 0;
+	}
+      port_start += np->port_block_size;
+      port_end = port_start + np->port_block_size;
+    }
+  while (port_end < np->max_port);
```

Comment:
I hate this trial and error approach to finding a suitable block

Reply:
I've removed this try-err thing out from UPG code and make logic for finding port block inside NAT code.
NAT plugin in full of such constructions in terms of port selection. Now its a part of binding validation procedure.